MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.c

```diff
@@ -0,0 +1,447 @@
+/**@file
+Library for defining PEI memory buckets. This keeps track of the
+different buckets and the data for the associated HOB.
+For internal use only.
+
+Copyright (c) Microsoft Corporation.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <PiPei.h>
+
+#include <Library/DebugLib.h>
+#include <Library/MemoryBucketLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/HobLib.h>
+
+// Memory types being kept in buckets in PEI.  Currently only runtime types.
+EFI_MEMORY_TYPE  MemoryTypes[4] = {
+  EfiRuntimeServicesCode,
+  EfiRuntimeServicesData,
+  EfiACPIReclaimMemory,
+  EfiACPIMemoryNVS
+};
+
+// PEI memory bucket statistics.  Can be extended if necessary
+EFI_MEMORY_TYPE_STATISTICS  RuntimeMemoryStats[4] = {
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiRuntimeServicesCode, TRUE, TRUE  },   // EfiRuntimeServicesCode
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiRuntimeServicesData, TRUE, TRUE  },   // EfiRuntimeServicesData
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiACPIReclaimMemory,   TRUE, FALSE },   // EfiACPIReclaimMemory
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiACPIMemoryNVS,       TRUE, FALSE }    // EfiACPIMemoryNVS
+};
+
+EFI_PHYSICAL_ADDRESS  CurrentBucketTops[4] = {
+  MAX_ALLOC_ADDRESS, // EfiRuntimeServicesCode
+  MAX_ALLOC_ADDRESS, // EfiRuntimeServicesData
+  MAX_ALLOC_ADDRESS, // EfiACPIReclaimMemory
+  MAX_ALLOC_ADDRESS  // EfiACPIMemoryNVS
+};
+
+// Total number of memory buckets
+UINTN  NumberOfBuckets = 4;
+
+// Information storage for Hob
+PEI_MEMORY_BUCKET_INFORMATION  RuntimeBucketHob;
+
+// TRUE if PEI memory buckets are disabled
+BOOLEAN  MemoryBucketsDisabled = FALSE;
+
+// TRUE if we have initialized the runtime memory buckets.
+BOOLEAN  RuntimeMemInitialized = FALSE;
+
+/**
+  TODO
+**/
+VOID
+EFIAPI
+InitializeMemoryBucketSizes (
+  VOID
+  )
+{
+  RuntimeMemoryStats[0].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketRuntimeCode);
+  RuntimeMemoryStats[1].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketRuntimeData);
+  RuntimeMemoryStats[2].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketAcpiReclaimMemory);
+  RuntimeMemoryStats[3].NumberOfPages = FixedPcdGet8 (PcdPeiMemoryBucketAcpiMemoryNvs);
+
+  // Disable memory buckets if the PCDs are unaltered.
+  if ((RuntimeMemoryStats[0].NumberOfPages + RuntimeMemoryStats[1].NumberOfPages +
+       RuntimeMemoryStats[2].NumberOfPages + RuntimeMemoryStats[3].NumberOfPages) == 0)
+  {
+    MemoryBucketsDisabled = TRUE;
+  }
+}
+
+/**
+  This function initialized the PEI memory buckets.  This should be called
+  when the first memory type being tracked in these buckets is allocated.
+
+  NOTE: The memory buckets can be created in Pre-mem PEI if runtime pages are
+  allocated in pre-mem PEI.

```

Comment:
Have you accounted for the pre-mem to post-mem memory transition in bucket addresses? If you're tracking addresses in your own HOB/area, you need to manually rebase your addresses across the transition.

Reply:
I added some logic for this but would appreciate feedback.

Reply:
I removed my initial logic and made it so that none of the PEI memory bucket stuff happens in pre-mem PEI.   It makes more sense to do this in my opinion.